org.eclipse.xtext/src/org/eclipse/xtext/generator/URIBasedFileSystemAccess.java

```diff
@@ -0,0 +1,244 @@
+/**
+ * Copyright (c) 2015, 2020 itemis AG (http://www.itemis.eu) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.xtext.generator;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.emf.ecore.resource.URIConverter;
+import org.eclipse.xtext.generator.trace.ITraceRegionProvider;
+import org.eclipse.xtext.generator.trace.TraceFileNameProvider;
+import org.eclipse.xtext.generator.trace.TraceNotFoundException;
+import org.eclipse.xtext.generator.trace.TraceRegionSerializer;
+import org.eclipse.xtext.parser.IEncodingProvider;
+import org.eclipse.xtext.util.RuntimeIOException;
+import org.eclipse.xtext.xbase.lib.CollectionLiterals;
+import org.eclipse.xtext.xbase.lib.Exceptions;
+import org.eclipse.xtext.xbase.lib.Pure;
+
+import com.google.common.io.ByteStreams;
+import com.google.common.io.CharStreams;
+
+/**
+ * A file system access implementation that is based on EMF URIs and URIConverter
+ * 
+ * @since 2.9
+ */
+public class URIBasedFileSystemAccess extends AbstractFileSystemAccess2 {
+	public interface BeforeDelete {
+		/**
+		 * @return <code>true</code> if the file can be deleted, false otherwise
+		 */
+		boolean beforeDelete(URI changed);
+	}
+
+	public interface BeforeWrite {
+		InputStream beforeWrite(URI changed, String outputCfgName, InputStream in);
+	}
+
+	public interface BeforeRead {
+		InputStream beforeRead(URI changed, InputStream in);
+	}
+
+	private URIConverter converter;
+	private URI baseDir;
+	private boolean generateTraces = false;
+	private IEncodingProvider encodingProvider = new IEncodingProvider.Runtime();
+	private TraceRegionSerializer traceRegionSerializer;
+	private TraceFileNameProvider traceFileNameProvider;
+	private URIBasedFileSystemAccess.BeforeDelete beforeDelete = (uri) -> true;
+	private URIBasedFileSystemAccess.BeforeWrite beforeWrite = (uri, string, stream) -> stream;
+	private URIBasedFileSystemAccess.BeforeRead beforeRead = (uri, stream) -> stream;
+
+	@Override
+	public void setPostProcessor(IFilePostProcessor filePostProcessor) {
+		super.setPostProcessor(filePostProcessor);
+	}
+
+	@Override
+	public URI getURI(String path, String outputConfiguration) {
+		String outlet = getPathes().get(outputConfiguration);
```

Comment:
API typos make me sad :(

Reply:
Nothing I can change here, right? It's protected.

Reply:
Nope, sadly you can't do anything about this